spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -152,84 +153,141 @@ public Subscriber subscribe(String subscription,
 	public PubsubMessage pullNext(String subscription) {
 		List<PubsubMessage> receivedMessageList = pullAndAck(subscription, 1, true);
 
-		return receivedMessageList.size() > 0 ?	receivedMessageList.get(0) : null;
+		return receivedMessageList.size() > 0 ? receivedMessageList.get(0) : null;
 	}
 
 	public SubscriberFactory getSubscriberFactory() {
 		return this.subscriberFactory;
 	}
 
 	@Override
-	public void ack(Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
-		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages can't be null.");
+	public ListenableFuture<Void> ack(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages) {
+		Assert.notEmpty(acknowledgeablePubsubMessages, "The acknowledgeablePubsubMessages cannot be null.");
 
-		groupAcknowledgeableMessages(acknowledgeablePubsubMessages).forEach(this::ack);
+		AcknowledgeablePubsubMessage acknowledgeablePubsubMessage = acknowledgeablePubsubMessages.iterator().next();
+
+		String projectId = acknowledgeablePubsubMessage.getProjectId();
+		String subscriptionName = acknowledgeablePubsubMessage.getSubscriptionName();
+		List<String> ackIds = collectAckIds(projectId, subscriptionName, acknowledgeablePubsubMessages);
+
+		SettableListenableFuture<Void> settableListenableFuture = new SettableListenableFuture<>();
+
+		ApiFuture<Empty> apiFuture = ack(subscriptionName, ackIds);
```

Comment:
Why do we worry about a `projectId` if we do nothing with it in the downstream logic?

Reply:
The `projectId` is used in `collectAckIds()` for validation that the `Collection< AcknowledgeablePubsubMessages>` all contain the same `projectId`.

Reply:
If the subscriptionId (name) is a String, then I think most developers would expect it to match the plain subscription id (name) when creating the subscription in GCP. (I'm using id (name) since it is inconsistent ... the API uses `id` ... the GCP UI uses `name`).

If we changed it to a `ProjectSubscriptionName`, then it would have the `projectId` and `subscriptionId` (name) ( `https://googleapis.github.io/googleapis/java/all/latest/apidocs/com/google/pubsub/v1/ProjectSubscriptionName.html`

Reply:
Sure, I think using `ProjectSubscriptionName` also makes sense.